### PR TITLE
trie: use Uint8Array as value type in DB

### PR DESCRIPTION
**Pull request by jochem-brouwer**
In this PR the trie will use values as Uint8Array in the DB.

TODOs

- [x] Ensure client uses Uint8Array by default
- [x] Add client flag to use old version of DB (which used strings)
- [x] Ensure VM and other places where tries are used (block?) an option to also use Uint8Array values there
- [x] Mark blockchain/state tests + client to use Uint8Array DB
- [x] Write tests on Trie

Note: I tested that with the new `--useStringValueTrieDB` flag, client can still run the "old version" of the DB. If this flag is not provided and an old DB version is ran, it fails.

This PR also:

- Ensures monorepo-wide we only create `new Trie` if we need it (we do not need a Trie when we want the transactions root of 0 transactions, for instance)
- Optimizes block `transactionsTrieIsValid`: calculate the transactionsRoot once and caches this (does not cache trie, just the root)

### CODE REVIEW

**Comment by holgerd77** [comment]
Really important here to check as early on as possible on the performance side. I guess this is really only with it if it brings at least a 3-5 % jump in overall block execution.

**Comment by holgerd77** [comment]
Updated this via UI

**Review by holgerd77 — APPROVED** [review]
Cool, synced the first 200.000 blocks with this, works like a charm! 🤩

Will merge.

File: packages/client/bin/cli.ts [hunk 1]
```diff
@@ -334,6 +334,14 @@ const args: ClientOpts = yargs(hideBin(process.argv))
...
       'Support for `--prefixStorageTrieKeys=false` is temporary. Please sync new instances with `prefixStorageTrieKeys` enabled',
   })
+  .options('useStringValueTrieDB', {
+    describe:
+      'Use string values in the trie DB. This is old behavior, new behavior uses Uint8Arrays in the DB (more performant)',
+    boolean: true,
+    default: false,
+    deprecated:
+      'Usage of old DBs which uses string-values is temporary. Please sync new instances without this option.',
+  })
```
[the remark this answers is not in the document]
Comment (jochem-brouwer) [reply packages/client/bin/cli.ts:344]: Ok, good point, so it is not being confused with a blocks DB or anything :smile:

**Pull request merged**